SubmitQueueEngine extension for arc land

sundargates opened the pull request:
- This is ready to be reviewed.
- We are building a 'quick' prototype and we will try submitting this for a formal review after that.

sundargates commented:
@Pawka Can you take a look at this review?

Pawka commented on src/land/ArcanistLandEngine.php:
```diff
@@ -86,6 +87,15 @@
     return $this;
   }
 
+  final public function getRevision() {
```
You can move these methods  ant `private $revision;` to `ArcanistSubmitQueueEngine` class. As they are used in mentioned engine only. So we'll leave this file untouched.

Pawka commented on src/configuration/ArcanistSettings.php:
```diff
@@ -188,6 +188,19 @@ private function getOptions() {
+      'uber.land.submitqueue.uri' => array(
+        'type' => 'string',
+        'help' => pht('URI to use for the submitqueue backend'),
+        'example' => '"https://submitqueue.uberinternal.com"',
```
Maybe default value would be useful here?

Pawka commented on src/configuration/ArcanistSettings.php:
```diff
@@ -188,6 +188,19 @@ private function getOptions() {
         'default' => false,
         'example' => 'false',
       ),
+      'uber.land.submitqueue.enable' => array(
```
IMO it would be enough to have `uber.land.submitqueue.uri` config only. If value is `null` or `false` - submit queue is disabled.

sundargates replied:
hmm I would actually prefer this as all the variables have one purpose.

Pawka commented on src/land/ArcanistSubmitQueueEngine.php:
```diff
@@ -0,0 +1,134 @@
+<?php
+
+final class ArcanistSubmitQueueEngine
```
Rename file and class to `UberArcanistSubmitQueueEngine`. We keep this agreement to identify which code parts were created by us.

Pawka commented on src/land/ArcanistSubmitQueueEngine.php:
```diff
@@ -0,0 +1,134 @@
+      }
+      $this->setRevision($this->getWorkflow()->getRevision());
+    }
+      // $this->writeInfo(
```
Cleanup.

Pawka commented on src/land/SubmitQueueClient.php:
```diff
@@ -0,0 +1,52 @@
+<?php
+
+final class SubmitQueueClient extends Phobject {
```
Rename file and class to `UberSubmitQueueClient`.

Pawka commented on src/land/SubmitQueueFuture.php:
```diff
@@ -0,0 +1,29 @@
+<?php
+
+final class SubmitQueueFuture extends FutureProxy {
```
Rename file and class to `UberSubmitQueueFuture`.

Pawka commented on src/workflow/ArcanistLandWorkflow.php:
```diff
@@ -29,6 +33,10 @@
   const REFTYPE_BOOKMARK = 'bookmark';
 
   public function getRevisionDict() {
+    if (!$this->revision) {
```
Are you sure this change will not break behavior?

Pawka commented on src/workflow/ArcanistLandWorkflow.php:
```diff
@@ -248,6 +256,14 @@ public function getArguments() {
           'actually modify or land the commits.'),
       ),
       '*' => 'branch',
+      'tbr' => array(
```
What `tbr` stands for?

sundargates replied:
to-be-reviewed. It's just the option name used at google/twitter.

Pawka commented on src/workflow/ArcanistLandWorkflow.php:
```diff
@@ -256,7 +272,13 @@ public function run() {
     $engine = null;
     if ($this->isGit && !$this->isGitSvn) {
-      $engine = new ArcanistGitLandEngine();
+      if ($this->shouldUseSubmitQueue && !$this->tbr) {
```
Avoid change existing code. Just add your your lines below:

```
$engine = new ArcanistGitLandEngine();
if ($this->shouldUseSubmitQueue && !$this->tbr) {
  $engine = new ArcanistSubmitQueueEngine(
  $this->submitQueueClient,
  $this->getConduit());
}
```

This is not "as effective" but might be easier merge upstream changes later.

Pawka commented on src/workflow/ArcanistLandWorkflow.php:
```diff
@@ -291,7 +313,8 @@ public function run() {
         ->setShouldPreview($this->preview)
-        ->setBuildMessageCallback(array($this, 'buildEngineMessage'));
+        ->setBuildMessageCallback(array($this, 'buildEngineMessage'))
+        ->setRevision($this->getRevisionDict());
```
Again, are you sure it will not break default behavior if revision is not set? Maybe it would be worth add check if submit queue is enabled:

```
if ($this->shouldUseSubmitQueue && !$this->tbr) {
  $engine->setRevision($this->getRevisionDict());
}
```

Pawka commented on src/workflow/ArcanistLandWorkflow.php:
```diff
@@ -682,7 +724,8 @@ private function printPendingCommits() {
   }
 
-  private function findRevision() {
+  public function getRevision() {
```
Do not rename existing methods. Avoid change existing code.

I even found `findRevision` medhot called at 344 line of this file.

Pawka commented on src/workflow/ArcanistLandWorkflow.php:
```diff
@@ -1612,16 +1659,20 @@ public function didCommitMerge() {
 
   public function didPush() {
-    $this->askForRepositoryUpdate();
+    if ($this->shouldUseSubmitQueue) {
```
Do not change existing code. Just add like this:
```
if ($this->shouldUseSubmitQueue) {
  return; // do nothing
}
<...>
```

Pawka commented:
- When modifying files which exists on upstream arcanist, try adding your code in such way as current code would be easily merged later.
- If you modify existing files, enclose your changes with comments. This will let us easily spot our changes later if merge conflict with upstream will happen.
```
// Start #UBER code
<your code here>
// End #UBER code
```
- Do not change existing behavior of methods. For example I'm not sure if `$this->getRevision()` call inside `getRevisionDict` will not make any problems later. Maybe add additional check "if submit queue is enabled and $this->revision is empty then...".
- And after all why you haven't considered create new workflow with unique name which would extend `ArcanistLandWorkflow`? So you could override required methods.

sundargates commented:
The reason a new workflow was not considered is because that essentially means you want to train your users with the new command. I would like this change to cause as little change/intrusion to the developers as possible. Plus, since we are using git, a lot of the land workflow has been transported to ArcanistGitLandEngine. For that, UberSubmitQueueEngine was created and we have tried to use as much methods from ArcanistGitLandEngine as possible.

sundargates commented:
@Pawka I have addressed all your comments. I think the overall change now is much more readable - thanks to your suggestions. Let me know if this is something that you wouldn't mind committing. 

Other questions that I have:
1. How does releasing new versions of arc work?
2. When SQ is enabled for a given repo, how do we ensure that users have the latest version of arc? (would custom arc unit engine that does `arc upgrade` work here?)
